flink-core/src/main/java/org/apache/flink/api/common/TaskInfo.java

```diff
@@ -107,4 +131,12 @@ public int getAttemptNumber() {
 	public String getTaskNameWithSubtasks() {
 		return this.taskNameWithSubtasks;
 	}
+
+	/**
+	 * Returns the allocation id for where this task is executed.
+	 * @return the allocation id for where this task is executed.
+	 */
+	public String getAllocationID() {
```

Comment:
I'm also not sure whether we want to expose location information to the `Task`. Given scheduling transparency, this feels a bit wrong. Isn't there a different way to check whether a task was recovered locally?

Reply:
The question is not only whether it is local but also if it has the same allocation id, so I don't think there is really a different way of figuring this out.

Reply:
Previously, I had a "hacky" version that extracted the allocation id from a to string method of the state store but I think stephan was more in favour of exposing this information through the context if needed. I cannot see how we can test if we have the same allocation id without having some way to access it.

Reply:
Hmm, what about testing it a bit more indirectly, by removing the checkpoint files on the DFS. Then you can only recover if you recover locally. Or by querying the REST interface? We might have to add the information to the `VertexTaskDetail`.

Otherwise, we start mixing concerns and expose unnecessary information to the user via the `AbstractRuntimeUDFContext`. Moreover, not every function has access to this information right now. For example the `RichAsyncFunctionRuntimeContext` does not expose it.

Reply:
What I find problematic about deleting the DFS files is that not all state is (yet) covered by local recovery and it is also a lot harder to debug the cause if there is an actual scheduling problem. With the current code, you can easily see which allocation was lost. REST API might be an option if it is somehow exposed there.

Reply:
How about this way: setup the cluster with one slot per TM, and use the `host:pid` to act as the "`allocationID`"?

Reply:
@sihuazhou that approach would not allow us to test that the scheduling works proper when having more than one slot per TM, which is a useful thing to test for. Furthermore, this also silently makes the assumption that there is a connection between allocation id and PID, but the test might want to also validate that this is actually true.